types/wlr_foreign_toplevel_management_v1.c

```diff
@@ -0,0 +1,583 @@
+	if (!resource) {
+		wl_client_post_no_memory(client);
+		return NULL;
+	}
+
+	wl_resource_set_implementation(resource, &toplevel_handle_impl, toplevel,
+	    foreign_toplevel_resource_destroy);
+
+	wl_list_insert(&toplevel->resources, wl_resource_get_link(resource));
+	zwlr_foreign_toplevel_manager_v1_send_toplevel(manager_resource, resource);
+	return resource;
+}
+
+struct wlr_foreign_toplevel_handle_v1 *
+wlr_foreign_toplevel_handle_v1_create(
+		struct wlr_foreign_toplevel_manager_v1 *manager) {
+	struct wlr_foreign_toplevel_handle_v1 *toplevel = calloc(1,
+			sizeof(struct wlr_foreign_toplevel_handle_v1));
+	if (!toplevel) {
+		return NULL;
+	}
+
+	wl_list_insert(&manager->toplevels, &toplevel->link);
+	toplevel->manager = manager;
+
+	wl_list_init(&toplevel->resources);
+	wl_list_init(&toplevel->outputs);
+
+	wl_signal_init(&toplevel->events.request_maximize);
+	wl_signal_init(&toplevel->events.request_minimize);
+	wl_signal_init(&toplevel->events.request_activate);
+	wl_signal_init(&toplevel->events.request_close);
+	wl_signal_init(&toplevel->events.set_rectangle);
+	wl_signal_init(&toplevel->events.destroy);
+
+	struct wl_resource *manager_resource, *tmp;
+	wl_resource_for_each_safe(manager_resource, tmp, &manager->resources) {
+		create_toplevel_resource_for_resource(toplevel, manager_resource);
+	}
+
+	return toplevel;
+}
+
+static const struct zwlr_foreign_toplevel_manager_v1_interface
+	foreign_toplevel_manager_impl;
+
+static void foreign_toplevel_manager_handle_stop(struct wl_client *client,
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwlr_foreign_toplevel_manager_v1_interface,
+		&foreign_toplevel_manager_impl));
+
+	zwlr_foreign_toplevel_manager_v1_send_finished(resource);
+	wl_resource_destroy(resource);
+}
+
+static const struct zwlr_foreign_toplevel_manager_v1_interface
+		foreign_toplevel_manager_impl = {
+	.stop = foreign_toplevel_manager_handle_stop
+};
+
+static void foreign_toplevel_manager_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static void toplevel_send_details_to_toplevel_resource(
+		struct wlr_foreign_toplevel_handle_v1 *toplevel,
+		struct wl_resource *resource) {
+	if (toplevel->title) {
+		zwlr_foreign_toplevel_handle_v1_send_title(resource,
+		    toplevel->title);
+	}
+	if (toplevel->app_id) {
+		zwlr_foreign_toplevel_handle_v1_send_app_id(resource,
+		    toplevel->app_id);
+	}
+
+	struct wlr_foreign_toplevel_handle_output *output;
+	wl_list_for_each(output, &toplevel->outputs, link) {
+		send_output_to_resource(resource, output->output, true);
+	}
+
+	struct wl_array states;
+	wl_array_init(&states);
+	bool r = fill_array_from_toplevel_state(&states, toplevel->state);
+	if (!r) {
+		wl_resource_post_no_memory(resource);
+		wl_array_release(&states);
+		return;
+	}
+
+	zwlr_foreign_toplevel_handle_v1_send_state(resource, &states);
+	wl_array_release(&states);
+
+	toplevel_update_idle_source(toplevel);
```

Comment:
This will send "done" to all resources -- we only want to send "done" to this one

Reply:
ok, but it would be ok if I directly send the done event to the resource without going through the idle loop in this case (because we already sent all we know about the toplevel)?

Reply:
Yeah